### ISSUE TITLE
[feature request] Show warning if optimizer.zero_grad() was not called before optimizer.step()

### ISSUE THREAD

**Issue by norpadon** [pytorch/pytorch#47175]
## Issue description

It is easy to introduce an extremely nasty bug in your code by forgetting to call `zero_grad()` or calling it at the beginning of each epoch instead of the beginning of each iteration.

I propose adding a simple check to ensure that `zero_grad()` is called before each `step()` and show a warning if it is violated. If this behavior is intentional (e.g., for gradient accumulation), one should be able to suppress the warning by passing an extra option to the `step()` function or the constructor of the optimizer

**Comment by kefirski** [comment]
It feels like such a design makes `zero_grad()` function to be useless at all. I would consider making zeroing gradients after calling the `step()` function to be default behaviour without calling `zero_grad()` with the ability to handle this operation like `step(zero_grad=False)` manually.

**Comment by norpadon** [comment]
@kefirski I think that it is slightly less intuitive and will silently break a lot of existing pipelines where gradient is accumulated over multiple mini-batches or multiple objectives are independently optimised at the same time. Warning is enough to eliminate all the bugs without breaking anything..

**Comment by mruberry** [comment]
Thank you for the recommendation. There's probably a better UX we could offer, but we would need to carefully review the details of any proposed design before accepting a change.

**Comment by norpadon** [comment]
I made this bloody mistake again this week and spent around 16 hours in total debugging it. I hate my life. Please add the damn warning.
